DfE.FindInformationAcademiesTrusts.Data.AcademiesDb/DataSourceProvider.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.EntityFrameworkCore;
+
+namespace DfE.FindInformationAcademiesTrusts.Data.AcademiesDb;
+
+public class DataSourceProvider : IDataSourceProvider
+{
+    private readonly IAcademiesDbContext _academiesDbContext;
+
+    [ExcludeFromCodeCoverage]
+    public DataSourceProvider(AcademiesDbContext academiesDbContext) : this(
+        (IAcademiesDbContext)academiesDbContext)
+    {
+    }
+
+    public DataSourceProvider(IAcademiesDbContext academiesDbContext)
+    {
+        _academiesDbContext = academiesDbContext;
+    }
+
+    public async Task<DataSource?> GetGiasUpdated()
+    {
+        var lastEntry = await _academiesDbContext.ApplicationEvents
+            .Where(e => e.Source != null
+                        && e.Source.Contains("adf-t1") && e.Source.Contains("-sips-dataflow")
+                        && e.Message == "Finished"
+                        && e.EventType != 'E'
+                        && e.Description == "GIAS_Daily").MaxAsync(e => e.DateTime);
+        if (lastEntry is null) return null;
```

Comment:
Rather than return null, we should log that we couldn't find the date and return a datasource with a null lastupdateddate